src/main/java/org/quiltmc/loader/impl/config/tree/TrackedValueImpl.java

```diff
@@ -0,0 +1,188 @@
+ * limitations under the License.
+ */
+
+package org.quiltmc.loader.impl.config.tree;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+
+import org.quiltmc.loader.api.config.Constraint;
+import org.quiltmc.loader.api.config.MetadataType;
+import org.quiltmc.loader.api.config.TrackedValue;
+import org.quiltmc.loader.api.config.values.ValueKey;
+import org.quiltmc.loader.impl.config.AbstractMetadataContainer;
+import org.quiltmc.loader.impl.config.ConfigImpl;
+import org.quiltmc.loader.impl.util.ImmutableIterable;
+
+public final class TrackedValueImpl<T> extends AbstractMetadataContainer implements TrackedValue<T> {
+	private final List<UpdateCallback<T>> callbacks;
+	private final List<Constraint<T>> constraints;
+	private final T defaultValue;
+
+	private ValueKey key;
+	private ConfigImpl config;
+	private T value;
+
+	private boolean isBeingOverridden = false;
+	private T valueOverride;
+
+	public TrackedValueImpl(ValueKey key, T defaultValue, Set<String> flags, Map<MetadataType<?>, List<?>> metadata, List<UpdateCallback<T>> callbacks, List<Constraint<T>> constraints) {
+		super(flags, metadata);
+		this.key = key;
+		this.defaultValue = defaultValue;
+		this.value = defaultValue;
+		this.callbacks = callbacks;
+		this.constraints = constraints;
+
+		this.assertValue(defaultValue);
+	}
+
+	public void setConfig(ConfigImpl config) {
+		if (this.config != null) {
+			throw new RuntimeException("TrackedValue '" + this.key + "' cannot be assigned to multiple configs");
+		}
+
+		this.config = config;
+	}
+
+	public TrackedValueImpl<T> setKey(ValueKey key) {
+		this.key = key;
+
+		return this;
+	}
+
+	@Override
+	public ValueKey getKey() {
+		return this.key;
+	}
+
+	@Override
+	public T getValue() {
+		return this.isBeingOverridden ? this.valueOverride : this.value;
+	}
+
+	@Override
+	public boolean isBeingOverridden() {
+		return this.isBeingOverridden;
+	}
+
+	@Override
+	public T getRealValue() {
+		return this.value;
+	}
+
+	private void assertValue(T value) {
+		Optional<Iterable<String>> errors = this.checkForFailingConstraints(value);
+
+		if (errors.isPresent()) {
+			StringBuilder errorMessage = new StringBuilder();
+
+			for (String message : errors.get()) {
+				errorMessage.append(message).append('\n');
+			}
+
+			throw new RuntimeException(errorMessage.toString());
+		}
+	}
+
+	@Override
+	public T setValue(T newValue, boolean serialize) {
+		this.assertValue(newValue);
+
+		T oldValue = this.value;
+		this.value = newValue;
+
+		if (serialize) {
+			this.config.serialize();
+		}
+
+		for (UpdateCallback<T> callback : this.callbacks) {
+			callback.onUpdate(this.key, oldValue, newValue);
+		}
+
+		return oldValue;
+	}
+
+	@Override
+	public void setOverride(T newValue) {
+		this.isBeingOverridden = true;
+		this.valueOverride = newValue;
+
+		for (UpdateCallback<T> callback : this.callbacks) {
+			callback.onUpdate(this.key, this.value, newValue);
```

Comment:
I don't see why not.

Reply:
Well, since you pass an oldValue and a newValue, I would assume the callback is invoked whenever `oldValue != newValue`.
Furthermore, I'd assume that oldValue was the previous value, which is also not quite true here.
I presonally think this should only trigger when the *visible* value (as returned by getValue) actually changed, but if you decide to *always* call it, I would probably just omit the `oldValue` and `newValue` arguments to avoid confusion.

Reply:
Just make sure to have a contract that you stick with and is consistent. Otherwise the API becomes hard to understand.
I'd rather have you *always* trigger the callbacks if any of these methods are called than sometimes, according to a rule that's not clearly visible.